internal/services/notification/runwebhookdelivery.go

```diff
@@ -0,0 +1,183 @@
+	"agola.io/agola/internal/sqlg/sql"
+	"agola.io/agola/internal/util"
+	"agola.io/agola/services/notification/types"
+)
+
+const (
+	MaxRunWebhookDeliveriesQueryLimit = 40
+	RunWebhookDeliveriesLockKey       = "runwebhookdeliveryevents"
+
+	// runWebhookDeliveriesInterval is the time to wait between every runWebhookDeliveriesHandler call.
+	runWebhookDeliveriesInterval = time.Second * 1
+)
+
+func (n *NotificationService) RunWebhookDeliveriesHandlerLoop(ctx context.Context) {
+	for {
+		if err := n.runWebhookDeliveriesHandler(ctx); err != nil {
+			n.log.Err(err).Send()
+		}
+
+		sleepCh := time.NewTimer(runWebhookDeliveriesInterval).C
+		select {
+		case <-ctx.Done():
+			return
+		case <-sleepCh:
+		}
+	}
+}
+
+func (n *NotificationService) runWebhookDeliveriesHandler(ctx context.Context) error {
+	l := n.lf.NewLock(RunWebhookDeliveriesLockKey)
+	if err := l.TryLock(ctx); err != nil {
+		if errors.Is(err, lock.ErrLocked) {
+			return nil
+		}
+		return errors.WithStack(err)
+	}
+	defer func() { _ = l.Unlock() }()
+
+	curRunWebhookDeliverySequence := uint64(0)
+
+	for {
+		var runWebhookDeliveries []*types.RunWebhookDelivery
+
+		err := n.d.Do(ctx, func(tx *sql.Tx) error {
+			var err error
+			runWebhookDeliveries, err = n.d.GetRunWebhookDeliveriesAfterSequence(tx, curRunWebhookDeliverySequence, types.DeliveryStatusNotDelivered, MaxRunWebhookDeliveriesQueryLimit)
+			if err != nil {
+				return errors.WithStack(err)
+			}
+
+			return nil
+		})
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		for _, r := range runWebhookDeliveries {
+			if err := n.handleRunWebhookDelivery(ctx, r); err != nil {
+				n.log.Err(err).Msgf("failed to trigger run webhook delivery event")
+			}
+
+			curRunWebhookDeliverySequence = r.Sequence
+		}
+
+		if len(runWebhookDeliveries) < MaxRunWebhookDeliveriesQueryLimit {
+			return nil
+		}
+	}
+}
+
+func (n *NotificationService) handleRunWebhookDelivery(ctx context.Context, runWebhookDelivery *types.RunWebhookDelivery) error {
+	var runWebhook *types.RunWebhook
+
+	err := n.d.Do(ctx, func(tx *sql.Tx) error {
+		var err error
+
+		runWebhook, err = n.d.GetRunWebhookByID(tx, runWebhookDelivery.RunWebhookID)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	webhookDelivered := false
+	resp, err := n.sendRunWebhook(ctx, runWebhook.Payload, runWebhook.ID)
+	// err != nil is not checked because every error is considered a failed delivery
+	if err == nil && resp != nil && resp.StatusCode == http.StatusCreated {
+		webhookDelivered = true
+	}
+
+	var deliveredAt *time.Time
+	var statusCode int
+
+	if resp != nil {
+		deliveredAt = util.TimeP(time.Now())
+		statusCode = resp.StatusCode
+	}
+
+	err = n.d.Do(ctx, func(tx *sql.Tx) error {
+		var err error
+
+		runWebhookDelivery, err = n.d.GetRunWebhookDeliveryByID(tx, runWebhookDelivery.ID)
+		if err != nil {
+			return errors.WithStack(err)
+		}
```

Comment:
get the runwebhookdelivery and check that it wasn't already delivered. In such case just do nothing. Remember, the lock is not mandatory.